content/articles/2019-02-purrr-0-3-0.Rmd

```diff
@@ -0,0 +1,308 @@
+---
+title: 'purrr 0.3.0'
+author: Lionel Henry
+date: '2019-02-04'
+slug: purrr-0-3-0
+description: >
+  purrr 0.3.0 is now on CRAN.
+categories:
+  - package
+photo:
+  url: https://unsplash.com/photos/NodtnCsLdTE
+  author: Mikhail Vasilyev
+---
+
+```{r setup, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>",
+  fig.width = 7,
+  fig.align = 'center',
+  fig.asp = 0.618, # 1 / phi
+  out.width = "700px"
+)
+
+library("purrr")
+```
+
+We're excited to announce the release of [purrr](https://purrr.tidyverse.org) 0.3.0! purrr enhances R’s functional programming toolkit by providing a complete and consistent set of tools for working with functions and vectors.
+
+```{r, eval = FALSE}
+# Install the latest version with:
+install.packages("purrr")
+
+# Start working with purrr:
+library("purrr")
+```
+
+In this release, `pluck()` gets a few new variants:
+
+* `pluck<-`, `assign_in()` and `modify_in()` allow deep modifications of nested structures.
+
+* `chuck()` is a stricter variant of `pluck()` that consistently fails instead of returning `NULL` when the structure does not have the expected shape.
+
+A new set of tools helps limiting the rate at which a function is called:
+
+* `slowly()` forces a function to sleep between each invokation.
+
+* `insistently()` automatically invokes a function again on error until it succeeds, and sleeps between invokations.
+
+* The `rate_delay()` and `rate_backoff()` helpers control the invokation rate of `slowly()` and `insistently()`.
+
+Finally, the consistency of the interface has been polished:
+
+* The direction of iteration/application is now consistently specified with a `.dir` argument.
+
+* Many missing functions were added to fill the gaps: `accumulate2()`, `imodify()`, `map_depth()`, ...
+
+* `partial()` has a much improved and more flexible interface.
+
+Find a detailed account of the changes in the [NEWS](https://github.com/tidyverse/purrr/blob/master/NEWS.md#purrr-030) file.
+
+
+## New pluck variants
+
+`pluck()` implements a generalised form of `[[` that allow you to index deeply and flexibly into data structures. For instance, `pluck(x, "foo", 2)` is equivalent to `x[["foo"]][[2]]`. purrr 0.3.0 introduces variants of `pluck()` to make it easier to work with deep data structures.
```

Comment:
Probably worth mentioning the `.default` argument here too: For instance, `pluck(x, "foo", 2, .default = NA)` is equivalent to `x[["foo"]][[2]]`, returning an `NA` if that element doesn't exist.